### PR TITLE
Add support for refunds

### PR DESCRIPTION
**Pull request by wwahammy**
Closed https://github.com/CommitChange/tix/issues/3632

Depends on #271

### CODE REVIEW

File: lib/insert/insert_refunds.rb [hunk 1]
```diff
@@ -22,9 +22,13 @@ def self.modern_refund(charge,h)
       nonprofit_id: {required: true, is_integer: true},
       supporter_id: {required: true, is_integer: true}
     })
+    if h['amount'].to_i != nil
+      ## TODO DO NOT MERGE AS IS
+      h['amount'] = h['amount'].to_i
+    end
```
Comment (clarissalimab) [inline lib/insert/insert_refunds.rb]: I think this means we can't merge it yet haha

File: lib/insert/insert_refunds.rb [hunk 2]
```diff
@@ -141,6 +150,7 @@ def self.legacy_refund(charge, h)
   # @option opts [Integer] :nonprofit_id the nonprofit_id that the charge belongs to
   # @option opts [Time] :charge_date the time that the charge to be refunded occurred
   def self.perform_stripe_refund(opts={})
+    byebug
```
Comment (clarissalimab) [inline lib/insert/insert_refunds.rb]:
```suggestion
```

File: app/models/concerns/model/subtransactable.rb [hunk 3]
```diff
@@ -13,8 +13,26 @@ module Model::Subtransactable
...
+			end
+	
+		end
 
 		delegate :currency, to: :nonprofit
+
+		# Handle a completed refund from a legacy Refund object
+		def process_refund(refund)
+			raise NotImplementedError
+		end
```
Comment (clarissalimab) [inline app/models/concerns/model/subtransactable.rb]: Do we want to keep that?

File: app/models/transaction.rb [hunk 4]
```diff
@@ -21,17 +21,27 @@ class Transaction < ApplicationRecord
...
+		new_refund = subtransaction.process_refund(refund)
+		self.amount = subtransaction.subtransactable.amount
+		transaction_assignments.process_refund(refund)
+		save!
+		publish_updated
+		subtransaction.publish_updated
+		ordered_payments.select{|i| i != new_refund}.each(&:publish_updated)
+		new_refund.publish_created
+		transaction_assignments.first.publish_updated
+	end
```
Comment (clarissalimab) [inline app/models/transaction.rb]: It's a bit hard to understand what's going on here, I would create methods for each part of it to make it clearer or add some comments.

**Review by clarissalimab — CHANGES_REQUESTED** [review]
Great work! Just a few comments :speech_balloon::

File: lib/insert/insert_refunds.rb (hunk 1 above)
Comment (wwahammy) [reply lib/insert/insert_refunds.rb]: Thank you!

File: app/models/transaction.rb (hunk 4 above)
Comment (wwahammy) [reply app/models/transaction.rb]: Fair point, let me think about how we can do that...

File: app/models/concerns/model/subtransactable.rb (hunk 3 above)
Comment (wwahammy) [reply app/models/concerns/model/subtransactable.rb]: Yes, the idea is that you need to override this in any subtransactable, like OfflineTransaction or StripeTransaction. That said, I should add a comment there to explain that.

Comment (wwahammy) [reply app/models/concerns/model/subtransactable.rb]: I've added an explanation to the error and documentation now

File: lib/insert/insert_refunds.rb (hunk 1 above)
Comment (wwahammy) [reply lib/insert/insert_refunds.rb]: But why didn't I want to merge as-is?

Comment (wwahammy) [reply lib/insert/insert_refunds.rb]: I've removed the warning, I don't see any reason why it's necessary.

File: app/models/transaction.rb (hunk 4 above)
Comment (wwahammy) [reply app/models/transaction.rb]: I've adjusted this in the latest and I think it makes it clearer.

**Review by clarissalimab — APPROVED** [review]
Nice work! :rocket:

**Pull request merged**